In [1]:
%%capture
%load_ext autoreload
%autoreload 2
from setup_nb_env import *

from epsampling.utils import load_csv

# pd.set_option('display.float_format', lambda x: '%.3f' % x)

data_dir = '/work/users/k/4/k4thryn/Repos/EpSampling/data/'

In [11]:
# timestamp = '20240920-125607'

# data_name = 'merged_covs_deaths'
# _df,name = load_csv(data_name, data_dir=f'{data_dir}processed/', timestamp=timestamp)
# _df.drop('Postal',axis=1,inplace=True)
# # print(name) 
# # display(_df)

# acs_covs = ['Fips'] + _df.columns[10:].values.tolist()
# df_acs = _df[acs_covs]
# df_acs.drop_duplicates(inplace=True)
# df_acs.dropna(inplace=True)

# death_covs = _df.columns[:10]
# df_deaths = _df[death_covs]

# d = datetime.today().strftime('%Y%m%d-%H%M%S')
# df_deaths.to_csv(f'{data_dir}cached/df_deaths_{d}.csv',index=False)
# df_acs.to_csv(f'{data_dir}cached/df_acs_{d}.csv',index=False)
# print(d)

20240925-112125


In [12]:
from epsampling.utils import load_csv

data_dir = '/work/users/k/4/k4thryn/Repos/EpSampling/data/'
timestamp = '20240925-112125'

df_deaths,_ = load_csv('df_deaths', data_dir=f'{data_dir}cached/', timestamp=timestamp)
df_acs,_ = load_csv('df_acs', data_dir=f'{data_dir}cached/', timestamp=timestamp)
# df_acs.set_index('Fips',inplace=True,drop=True)

display(df_deaths, df_acs)

State_fips State          County   Fips        Date   
0               41    OR           Baker  41001  2020-05-09  \
1               41    OR           Baker  41001  2020-05-16   
2               41    OR           Baker  41001  2020-05-23   
3               41    OR           Baker  41001  2020-05-30   
4               41    OR           Baker  41001  2020-06-06   
...            ...   ...             ...    ...         ...   
332432          24    MD  Baltimore city  24510  2022-04-09   
332433          24    MD  Baltimore city  24510  2022-04-16   
332434          24    MD  Baltimore city  24510  2022-04-23   
332435          24    MD  Baltimore city  24510  2022-04-30   
332436          24    MD  Baltimore city  24510  2022-05-07   

        COVIDhubEns_state_deaths     Pop  Pop_ratio  True_county_deaths   
0                        125.655   16124   0.003823                 0.0  \
1                        143.950   16124   0.003823                 0.0   
2                        150.378   16124   0.003823                 0.0   
3                        159.093   16124   0.003823                 0.0   
4                        164.199   16124   0.003823                 0.0   
...                          ...     ...        ...                 ...   
332432                 14367.000  593490   0.098168              1765.0   
332433                 14391.000  593490   0.098168              1771.0   
332434                 14423.000  593490   0.098168              1771.0   
332435                 14443.000  593490   0.098168              1773.0   
332436                 14488.000  593490   0.098168              1776.0   

        Naive_county_deaths  
0                  0.480367  
1                  0.550307  
2                  0.574881  
3                  0.608197  
4                  0.627717  
...                     ...  
332432          1410.374156  
332433          1412.730179  
332434          1415.871543  
332435          1417.834895  
332436          1422.252438  

[332437 rows x 10 columns]

Fips  POP_x2   POP_M   POP_F  POP_A0004  POP_A0509  POP_A1014   
0     41001   16539    8458    8081        855       1018        939  \
1     41003   94667   47675   46992       3735       3837       4861   
2     41005  418577  207731  210846      21760      25709      26230   
3     41007   40720   20226   20494       1896       1802       2494   
4     41009   52381   26374   26007       2799       3242       2886   
...     ...     ...     ...     ...        ...        ...        ...   
3124   8119   24607   12517   12090        874       1377       1161   
3125   8121    4834    2573    2261        213        350        318   
3126   8123  322424  163656  158768      22572      23842      23870   
3127   8125    9944    5059    4885        725        812        794   
3128  24510  592211  278792  313419      36513      32613      34006   

      POP_A1517  POP_A1819  POP_A20  ...  HU_x15  HU_UIS01D  HU_UIS01A   
0           561        160       99  ...    8654       6300        109  \
1          2958       6348     3078  ...   39636      23185       2308   
2         16532       8742     4512  ...  169113     116490       7926   
3          1382        855      698  ...   22882      15712        491   
4          2038       1111      457  ...   21586      15908        370   
...         ...        ...      ...  ...     ...        ...        ...   
3124        728        365      132  ...   13271      11115        263   
3125        200         94       40  ...    2287       1942         41   
3126      13892       9715     4363  ...  115854      84827       4869   
3127        396        196       59  ...    4311       3523        121   
3128      18424      16653     8764  ...  293718      43042     149324   

      HU_UIS02  HU_UIS0304  HU_UIS0509  HU_UIS1019  HU_UIS2049  HU_UIS50P   
0          162         159         183         105          90        157  \
1         1002        2615        3001        1542        1577       2045   
2         2094        5293        7750        5969        4692       8914   
3         1283        1341         752         461         820        556   
4          754         622         260         465         110        287   
...        ...         ...         ...         ...         ...        ...   
3124        70         200         223         239         125         35   
3125        31          35          12           8          19          4   
3126      2184        3683        4283        3445        2492       2123   
3127        17          78          71           6          30          4   
3128     14615       16871       17300       12674        7999      31436   

      HU_UISOTHER  
0            1389  
1            2361  
2            9985  
3            1466  
4            2810  
...           ...  
3124         1001  
3125          195  
3126         7948  
3127          461  
3128          457  

[3129 rows x 192 columns]

### <font color=blue> Feature selection: Correlation.

In [13]:
## Remove highly correlated features.
corr_matrix = df_acs.corr().abs()

## Select upper triangle of correlation matrix
upper = corr_matrix.where(np.triu(np.ones(corr_matrix.shape), k=1).astype(bool))

## Find features with correlation greater than 0.95
to_drop = [column for column in upper.columns if any(upper[column] > 0.95)]

## Drop features 
df_acs_filt = df_acs.drop(to_drop, axis=1, inplace=False)
df_acs_filt

Fips  POP_x2  EDU_ltHS      MHI  IND_AFFHM  IND_MAN_x7  IND_INF_x7   
0     41001   16539      1130  46922.0        639         591         136  \
1     41003   94667      2143  68732.0       1444        4122         697   
2     41005  418577     16842  88517.0       4079       22391        3006   
3     41007   40720      2361  61846.0        620        1835         211   
4     41009   52381      3606  73909.0        578        4249          84   
...     ...     ...       ...      ...        ...         ...         ...   
3124   8119   24607       414  68677.0        516         663         255   
3125   8121    4834       290  54141.0        470         113          33   
3126   8123  322424     24878  80843.0       8942       16918        2251   
3127   8125    9944       819  56327.0        975         127          56   
3128  24510  592211     56716  54124.0        638       12518        4776   

      IND_PUBA_x7  IND_AFFHM_AFFH  IND_AFFHM_MQE  ...  POP_ASIANNH   
0             472             609             30  ...           58  \
1            1834            1425             19  ...         6905   
2            8131            3828            251  ...        18777   
3             796             561             59  ...          416   
4            1157             490             88  ...          401   
...           ...             ...            ...  ...          ...   
3124          517             287            229  ...          139   
3125          233             453             17  ...           18   
3126         6162            4079           4863  ...         5266   
3127          195             914             61  ...            0   
3128        21571             493            145  ...        14763   

      POP_NHPINH  POP_OTH2PLNH  POP_AIAN  HU_UIS01A  HU_UIS02  HU_UIS0304   
0             32           412       235        109       162         159  \
1            222          3722       691       2308      1002        2615   
2            817         21432      2193       7926      2094        5293   
3             95          2209       297        491      1283        1341   
4            190          2837       440        370       754         622   
...          ...           ...       ...        ...       ...         ...   
3124           4          1311       161        263        70         200   
3125           1            72         0         41        31          35   
3126         348          8146      2236       4869      2184        3683   
3127           0           104        53        121        17          78   
3128         131         19989      1706     149324     14615       16871   

      HU_UIS2049  HU_UIS50P  HU_UISOTHER  
0             90        157         1389  
1           1577       2045         2361  
2           4692       8914         9985  
3            820        556         1466  
4            110        287         2810  
...          ...        ...          ...  
3124         125         35         1001  
3125          19          4          195  
3126        2492       2123         7948  
3127          30          4          461  
3128        7999      31436          457  

[3129 rows x 24 columns]

In [14]:
from epsampling.utils import time_to_int
df_deaths['Timestring'] = df_deaths.Date.apply(lambda x: time_to_int(x))

In [26]:
display(df_deaths, df_acs_filt)

State_fips State          County   Fips        Date   
0               41    OR           Baker  41001  2020-05-09  \
1               41    OR           Baker  41001  2020-05-16   
2               41    OR           Baker  41001  2020-05-23   
3               41    OR           Baker  41001  2020-05-30   
4               41    OR           Baker  41001  2020-06-06   
...            ...   ...             ...    ...         ...   
332432          24    MD  Baltimore city  24510  2022-04-09   
332433          24    MD  Baltimore city  24510  2022-04-16   
332434          24    MD  Baltimore city  24510  2022-04-23   
332435          24    MD  Baltimore city  24510  2022-04-30   
332436          24    MD  Baltimore city  24510  2022-05-07   

        COVIDhubEns_state_deaths     Pop  Pop_ratio  True_county_deaths   
0                        125.655   16124   0.003823                 0.0  \
1                        143.950   16124   0.003823                 0.0   
2                        150.378   16124   0.003823                 0.0   
3                        159.093   16124   0.003823                 0.0   
4                        164.199   16124   0.003823                 0.0   
...                          ...     ...        ...                 ...   
332432                 14367.000  593490   0.098168              1765.0   
332433                 14391.000  593490   0.098168              1771.0   
332434                 14423.000  593490   0.098168              1771.0   
332435                 14443.000  593490   0.098168              1773.0   
332436                 14488.000  593490   0.098168              1776.0   

        Naive_county_deaths  Timestring  
0                  0.480367     4334940  
1                  0.550307     4335360  
2                  0.574881     4335780  
3                  0.608197     4336200  
4                  0.627717     4338360  
...                     ...         ...  
332432          1410.374156     4504140  
332433          1412.730179     4504560  
332434          1415.871543     4504980  
332435          1417.834895     4505400  
332436          1422.252438     4507620  

[332437 rows x 11 columns]

Fips  POP_x2  EDU_ltHS      MHI  IND_AFFHM  IND_MAN_x7  IND_INF_x7   
0     41001   16539      1130  46922.0        639         591         136  \
1     41003   94667      2143  68732.0       1444        4122         697   
2     41005  418577     16842  88517.0       4079       22391        3006   
3     41007   40720      2361  61846.0        620        1835         211   
4     41009   52381      3606  73909.0        578        4249          84   
...     ...     ...       ...      ...        ...         ...         ...   
3124   8119   24607       414  68677.0        516         663         255   
3125   8121    4834       290  54141.0        470         113          33   
3126   8123  322424     24878  80843.0       8942       16918        2251   
3127   8125    9944       819  56327.0        975         127          56   
3128  24510  592211     56716  54124.0        638       12518        4776   

      IND_PUBA_x7  IND_AFFHM_AFFH  IND_AFFHM_MQE  ...  POP_ASIANNH   
0             472             609             30  ...           58  \
1            1834            1425             19  ...         6905   
2            8131            3828            251  ...        18777   
3             796             561             59  ...          416   
4            1157             490             88  ...          401   
...           ...             ...            ...  ...          ...   
3124          517             287            229  ...          139   
3125          233             453             17  ...           18   
3126         6162            4079           4863  ...         5266   
3127          195             914             61  ...            0   
3128        21571             493            145  ...        14763   

      POP_NHPINH  POP_OTH2PLNH  POP_AIAN  HU_UIS01A  HU_UIS02  HU_UIS0304   
0             32           412       235        109       162         159  \
1            222          3722       691       2308      1002        2615   
2            817         21432      2193       7926      2094        5293   
3             95          2209       297        491      1283        1341   
4            190          2837       440        370       754         622   
...          ...           ...       ...        ...       ...         ...   
3124           4          1311       161        263        70         200   
3125           1            72         0         41        31          35   
3126         348          8146      2236       4869      2184        3683   
3127           0           104        53        121        17          78   
3128         131         19989      1706     149324     14615       16871   

      HU_UIS2049  HU_UIS50P  HU_UISOTHER  
0             90        157         1389  
1           1577       2045         2361  
2           4692       8914         9985  
3            820        556         1466  
4            110        287         2810  
...          ...        ...          ...  
3124         125         35         1001  
3125          19          4          195  
3126        2492       2123         7948  
3127          30          4          461  
3128        7999      31436          457  

[3129 rows x 24 columns]

In [15]:
df = df_deaths.merge(df_acs_filt,on='Fips')
df

State_fips State          County   Fips        Date   
0               41    OR           Baker  41001  2020-05-09  \
1               41    OR           Baker  41001  2020-05-16   
2               41    OR           Baker  41001  2020-05-23   
3               41    OR           Baker  41001  2020-05-30   
4               41    OR           Baker  41001  2020-06-06   
...            ...   ...             ...    ...         ...   
332355          24    MD  Baltimore city  24510  2022-04-09   
332356          24    MD  Baltimore city  24510  2022-04-16   
332357          24    MD  Baltimore city  24510  2022-04-23   
332358          24    MD  Baltimore city  24510  2022-04-30   
332359          24    MD  Baltimore city  24510  2022-05-07   

        COVIDhubEns_state_deaths     Pop  Pop_ratio  True_county_deaths   
0                        125.655   16124   0.003823                 0.0  \
1                        143.950   16124   0.003823                 0.0   
2                        150.378   16124   0.003823                 0.0   
3                        159.093   16124   0.003823                 0.0   
4                        164.199   16124   0.003823                 0.0   
...                          ...     ...        ...                 ...   
332355                 14367.000  593490   0.098168              1765.0   
332356                 14391.000  593490   0.098168              1771.0   
332357                 14423.000  593490   0.098168              1771.0   
332358                 14443.000  593490   0.098168              1773.0   
332359                 14488.000  593490   0.098168              1776.0   

        Naive_county_deaths  ...  POP_ASIANNH  POP_NHPINH  POP_OTH2PLNH   
0                  0.480367  ...           58          32           412  \
1                  0.550307  ...           58          32           412   
2                  0.574881  ...           58          32           412   
3                  0.608197  ...           58          32           412   
4                  0.627717  ...           58          32           412   
...                     ...  ...          ...         ...           ...   
332355          1410.374156  ...        14763         131         19989   
332356          1412.730179  ...        14763         131         19989   
332357          1415.871543  ...        14763         131         19989   
332358          1417.834895  ...        14763         131         19989   
332359          1422.252438  ...        14763         131         19989   

        POP_AIAN  HU_UIS01A  HU_UIS02  HU_UIS0304  HU_UIS2049  HU_UIS50P   
0            235        109       162         159          90        157  \
1            235        109       162         159          90        157   
2            235        109       162         159          90        157   
3            235        109       162         159          90        157   
4            235        109       162         159          90        157   
...          ...        ...       ...         ...         ...        ...   
332355      1706     149324     14615       16871        7999      31436   
332356      1706     149324     14615       16871        7999      31436   
332357      1706     149324     14615       16871        7999      31436   
332358      1706     149324     14615       16871        7999      31436   
332359      1706     149324     14615       16871        7999      31436   

        HU_UISOTHER  
0              1389  
1              1389  
2              1389  
3              1389  
4              1389  
...             ...  
332355          457  
332356          457  
332357          457  
332358          457  
332359          457  

[332360 rows x 34 columns]

In [22]:
# # # # # # # # # # # # 
test_size = 0.1
# # # # # # # # # # # # 

## Date split (extrapolate)
# dates = sorted(df.Date.unique())
# lim = int(len(dates) * test_size)
# dates_test = dates[lim:]
# dates_train = dates[:lim]
# # print(len(dates_test), len(dates_train))
# df_train = df[df.Date.isin(dates_train)]
# df_test = df[df.Date.isin(dates_test)]

## Random split
from sklearn.model_selection import train_test_split
df_train, df_test = train_test_split(df, test_size=test_size, random_state=666)



In [23]:
death_covs = ['COVIDhubEns_state_deaths', 'Pop', 'Pop_ratio', 'Naive_county_deaths', 'Timestring']
feats = death_covs + df_acs_filt.columns.values.tolist()

X_train = df_train[feats]
X_test = df_test[feats]

y_train = df_train['True_county_deaths']
y_test = df_test['True_county_deaths']
y_naive = df_test['Naive_county_deaths']

In [24]:
from sklearn.ensemble import RandomForestRegressor
# train time for random split at 0.1 test_size? 2 min
clf = RandomForestRegressor(random_state=666, max_depth=6)
clf.fit(X_train, y_train)

RandomForestRegressor(max_depth=6, random_state=666)

In [25]:
from epsampling.utils import get_performance

y_pred = clf.predict(X_test)

model_names = ['Deaths_pred', 'Deaths_naive']
model_preds = [y_pred, y_naive]

metrics_dict = get_performance(model_names, model_preds, y_test, y_naive)
metrics_dict

{'Deaths_pred': {'MAE': 36.394611487325754,
  'MSE': 8381.72489102454,
  'r2': 0.9823884676266588,
  'relMAE': 4.339679264994637},
 'Deaths_naive': {'MAE': 47.10766213002411,
  'MSE': 40221.25637260506,
  'r2': 0.9154878061601462,
  'relMAE': 1.0}}

{
'Time_split':  
 {'MAE': 95.46361067933773,
  'MSE': 204383.6363671881,
  'r2': 0.5192914234530379,
  'relMAE': 10.398700987761641},

'Random_split': 
 {'MAE': 36.394611487325754,
  'MSE': 8381.72489102454,
  'r2': 0.9823884676266588,
  'relMAE': 4.339679264994637}

'Deaths_naive': 
 {'MAE': 48.00317963142656,
  'MSE': 34997.352296998506,
  'r2': 0.9176865246913497,
  'relMAE': 1.0}
  }


















### <font color=magenta> Conclusion: ... Random split does better than time split (extrapolation), but does not beat naive method.
## <font color=mediumspringgreen> NEXT: Try adding previous true death counts to X features.

In [28]:
# display(df_deaths, df_acs_filt)
from datetime import datetime

d = datetime.today().strftime('%Y%m%d-%H%M%S')
df_deaths.to_csv(f'{data_dir}cached/df_deaths_filt_{d}.csv',index=False)
df_acs_filt.to_csv(f'{data_dir}cached/df_acs_filt_{d}.csv',index=False)
print(d)

20240925-113849
